File: my-folder/0084-largest-rectangle-in-histogram/solution.cpp
class Solution {
public:
    int max(int a,int b)
    {
        if(a<b)
            return b;
        return a;
    }
    int largestRectangleArea(vector<int>& arr) {
        int n=arr.size();
        int res=0,curr;
        stack<int>s;
        for(int i=0;i<n;i++)
        {
            while(!s.empty() && arr[s.top()]>=arr[i])
            {
                int t=s.top();
                s.pop();
                curr=arr[t]*(s.empty()?i:i-s.top()-1);
                res=max(res,curr);
            }
            s.push(i);
        }
        while(!s.empty())
        {
            int t=s.top();
            s.pop();
            curr=arr[t]*(s.empty()?n:n-s.top()-1);
            res=max(res,curr);
        }
        return res; 
    }
};

File: my-folder/0084-largest-rectangle-in-histogram/solution.java
class Solution {
    public int largestRectangleArea(int[] heights) {

        int n = heights.length;
        int next[] = new int[n];
        int prev[] = new int[n];
        Stack<Integer> st = new Stack<>();

        // Next smaller
        for(int i=n-1;i>=0;i--){
            while(!st.isEmpty() && heights[st.peek()]>=heights[i])
                st.pop();

            next[i] = st.isEmpty() ? n : st.peek();
            st.push(i);
        }

        st.clear();

        // Previous smaller
        for(int i=0;i<n;i++){
            while(!st.isEmpty() && heights[st.peek()]>=heights[i])
                st.pop();

            prev[i] = st.isEmpty() ? -1 : st.peek();
            st.push(i);
        }

        int max=0;

        for(int i=0;i<n;i++){
            int width = next[i]-prev[i]-1;
            max = Math.max(max, heights[i]*width);
        }

        return max;
    }
}
